app/components/current-page.ts

```diff
@@ -0,0 +1,99 @@
+import type Store from '@ember-data/store';
```

Comment:
here is what this would look like once I fix a bug with `trackedFunction`:

```ts
import type Store from '@ember-data/store';
import type Book from 'collection-demo/models/book';
import type { Collection } from '@ember-data/types';
import { resource, resourceFactory } from 'ember-resources';

interface FirstPageOptions {
  url: string;
  sort?: string | null;
  filter?: string | null;
  genre?: string | null;
  author?: string | null;
  page?: number | null;
  limit?: number | null;
}

export const CurrentPage = resourceFactory(
  /**
   * This way of passing args has nothing to do with resources,
   * but is just how you pass args for lazy evaluation without creating
   * an options class with its own tracked properties
   */
  (optionsFns: {
    currentUrl: () => string | null;
    firstPageOptions: () => FirstPageOptions;
    addPage: (content: Collection<unknown>) => void;
  }) => {
    let previousFirstPageOptions: FirstPageOptions | undefined;

    return resource(({ on, owner, use }) => {
      const dataWrapper = new AsyncContent<Collection<Book>>();
      const addPage = optionsFns.addPage;

      const store = owner.lookup('service:store') as unknown as Store;

      const request = use(TrackedFunction(async () => {
        const currentUrl = optionsFns.currentUrl();
        const firstPageOptions = optionsFns.firstPageOptions();

        let options = firstPageOptions;

        if (previousFirstPageOptions?.url === firstPageOptions.url) {
          options = currentUrl ? { url: currentUrl } : firstPageOptions;
        }

        const books = await store.request<Book>(options);

        return books.content;
      }));

  
      return request;
    });
  }
);
```